### DISCUSSION TITLE
advance install instructions not working

### DISCUSSION THREAD

**Discussion by ekrocks** [bullhorn/career-portal#425] [Q&A] [answered]
I have been following instructions on the advanced install to enable google job integration with sitemap. I noticed that there are few instructions that are bit unclear. 

I used https://github.com/bullhorn/career-portal/blob/master/docs/advanced-install.md to install and follow the instructions and I was wondering if I can get some clarifications.
- I noticed that there isn't /server folder when downloading the the most recent release .zip. They are only created when building the cloned code using ```npm run build``` under /dist/server. Is this the correct folder that is mentioned in the doc?
- The instruction says to use ```pm2 start server.js``` when in dist/server folder but the server.js is in /dist folder rather than in /dist/server. Am I suppose to move the file to the server folder?
- When I run the server.js using pm2, I get CONNECTION_REFUSED error while I am able to build the project properly onto ```localhost:4000``` when using ```npm start```. Would this error be connected to anything I have mentioned above?
- With regards to sitemap documentation, when running pm2 start command, is it suppose to create the sitemap.xml file? Is this file supposed to be created when using ```npm run build``` as well?

EDIT:
- to make my PM2 run locally, I had to move the server.js file to the root folder. However when I try to get the sitemap using ```rootadress/sitemap```, I get this error in the PM2 Log
 ```
server > Error: connect ETIMEDOUT IPADDRESS:443                                                               
server >     at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1148:16) {                                    
server >   errno: -4039,                                                                              
server >   code: 'ETIMEDOUT',                                                                                     
server >   syscall: 'connect',                                                                                    
 server >   address: 'IPREMOVED',                                                                              
 server >   port: 443                                                                                              
 server > }
```
Thank you

**Comment by charlesabarnes** [comment, marked as answer]
Hello,

I believe I need to revamp the advanced documentation because I have excluded the advanced installation package.   That is the correct folder that is generated and I advise using the `npm run build` command for it to generate the correct files.

the pm2 start command would need to run in the dist folder, that's correct.   If you run `pm2 logs` to find out additional information out about the error you are receiving.

The sitemap is not an actual file.  Its generated when you go to the `/sitemap` path on your site

**Reply by ekrocks** [reply]
wow, thank you for a quick reply again, I have updated the question with the errors. 

I have used ```run build``` on the latest code changes and still the problem persisted.

When I was running ```pm2 start``` on /dist folder I was getting this: 
```0|server   |     at Object.openSync (fs.js:498:3)
0|server   |     at readFileSync (fs.js:394:35)
0|server   |     at Module.<anonymous> (C:\Users\USERNAME\source\reposVSC\Protrans Careers\dist\server.js:128:71)
0|server   |     at __webpack_require__ (C:\Users\USERNAME\source\reposVSC\Protrans Careers\dist\server.js:20:30)
0|server   |     at C:\Users\USERNAME\source\reposVSC\Protrans Careers\dist\server.js:84:18
0|server   |     at Object.<anonymous> (C:\Users\USERNAME\source\reposVSC\Protrans Careers\dist\server.js:87:10)
0|server   |     at Module._compile (internal/modules/cjs/loader.js:1072:14)
0|server   |     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10)
0|server   |     at Module.load (internal/modules/cjs/loader.js:937:32)
0|server   |     at Function.Module._load (internal/modules/cjs/loader.js:778:12) {
0|server   |   errno: -4058,
0|server   |   syscall: 'open',
0|server   |   code: 'ENOENT',
0|server   |   path: 'C:\\Users\\USERNAME\\source\\reposVSC\\Protrans Careers\\dist\\dist\\browser\\index.html'
0|server   | }
``` 

I noticed the duplicated folder on the last line which made me move the ```server.js``` to the root folder. 

Regarding with ```/sitemap``` I updated the discussion regarding with the error I get as well

**Reply by charlesabarnes** [reply]
Good call on moving the file.  Just so I understand, is the main site working now (with the moved ,js file), and its just the sitemap not working?

**Reply by charlesabarnes** [reply]
Oh, also, did you update the `app.json` file in the dist directory with the requisite information (corpToken/swimlane)?

**Reply by ekrocks** [reply]
Yes when I moved it to root folder 
![image](https://user-images.githubusercontent.com/10538258/134566925-d299cde6-4b8f-4120-9714-c3100f9cf02e.png)
and it shows the portal properly.

Sitemap seems to be the only thing not working now.

**Reply by ekrocks** [reply]
I noticed that there are 2 app.json files in one in `/dist/browser/app.json` and `/src/app.json` which I have put the proper swimlane and token. 

When I start the server through pm2 and as soon as I try to connect it I also get this error
```
ERROR HttpErrorResponse {
0|server  |   headers: HttpHeaders {
0|server  |     normalizedNames: Map(0) {},
0|server  |     lazyUpdate: null,
0|server  |     headers: Map(0) {}
0|server  |   },
0|server  |   status: 0,
0|server  |   statusText: 'Unknown Error',
0|server  |   url: 'https://public-rest{hereisswimlane}.bullhornstaffing.com:443/rest-services/{tokenhere}/search/JobOrder?query=(isOpen:1) AND (isDeleted:0)&count=500&fields=id&sort=id',
0|server  |   ok: false,
0|server  |   name: 'HttpErrorResponse',
0|server  |   message: 'Http failure response for https://public-rest{hereisswimlane}.bullhornstaffing.com:443/rest-services/{tokenhere}/search/JobOrder?query=(isOpen:1) AND (isDeleted:0)&count=500&fields=id&sort=id: 0 Unknown Error',
0|server  |   error: ProgressEvent {
0|server  |     type: 'error',
0|server  |     target: XMLHttpRequest {
0|server  |       onloadstart: null,
0|server  |       onprogress: null,
0|server  |       onabort: null,
0|server  |       onerror: null,
0|server  |       onload: null,
0|server  |       ontimeout: null,
0|server  |       onloadend: null,
0|server  |       _listeners: [Object],
0|server  |       onreadystatechange: null,
0|server  |       _anonymous: undefined,
0|server  |       readyState: 4,
0|server  |       response: null,
0|server  |       responseText: '',
0|server  |       responseType: 'text',
0|server  |       responseURL: '',
0|server  |       status: 0,
0|server  |       statusText: '',
0|server  |       timeout: 0,
0|server  |       upload: [XMLHttpRequestUpload],
0|server  |       _method: 'GET',
0|server  |       _url: [Url],
0|server  |       _sync: false,
0|server  |       _headers: [Object],
0|server  |       _loweredHeaders: [Object],
0|server  |       _mimeOverride: null,
0|server  |       _request: null,
0|server  |       _response: null,
0|server  |       _responseParts: null,
0|server  |       _responseHeaders: null,
0|server  |       _aborting: null,
0|server  |       _error: null,
0|server  |       _loadedBytes: 0,
0|server  |       _totalBytes: 0,
0|server  |       _lengthComputable: false
0|server  |     },
0|server  |     currentTarget: XMLHttpRequest {
0|server  |       onloadstart: null,
0|server  |       onprogress: null,
0|server  |       onabort: null,
0|server  |       onerror: null,
0|server  |       onload: null,
0|server  |       ontimeout: null,
0|server  |       onloadend: null,
0|server  |       _listeners: [Object],
0|server  |       onreadystatechange: null,
0|server  |       _anonymous: undefined,
0|server  |       readyState: 4,
0|server  |       response: null,
0|server  |       responseText: '',
0|server  |       responseType: 'text',
0|server  |       responseURL: '',
0|server  |       status: 0,
0|server  |       statusText: '',
0|server  |       timeout: 0,
0|server  |       upload: [XMLHttpRequestUpload],
0|server  |       _method: 'GET',
0|server  |       _url: [Url],
0|server  |       _sync: false,
0|server  |       _headers: [Object],
0|server  |       _loweredHeaders: [Object],
0|server  |       _mimeOverride: null,
0|server  |       _request: null,
0|server  |       _response: null,
0|server  |       _responseParts: null,
0|server  |       _responseHeaders: null,
0|server  |       _aborting: null,
0|server  |       _error: null,
0|server  |       _loadedBytes: 0,
0|server  |       _totalBytes: 0,
0|server  |       _lengthComputable: false
0|server  |     },
0|server  |     lengthComputable: false,
0|server  |     loaded: 0,
0|server  |     total: 0
0|server  |   }
0|server  | }
```
I get bunch of these errors in the pm2 log. Not sure if its related to sitemap error.

**Reply by charlesabarnes** [reply]
What do you get if you navigate to the URI directly?

**Reply by charlesabarnes** [reply]
the uri in the message I mean

**Reply by ekrocks** [reply]
I seem to get what looks like a JSON file:
![image](https://user-images.githubusercontent.com/10538258/134667955-6e0fab98-d95c-405c-9d22-09bfb64133d0.png)

Is this what we want?

This error is quite odd as it has no problem viewing the website even with the errors showing every time I restart the local instance. 

Also is PM2 Required for the advanced Hosting? Are there any alternatives other than PM2? Our servers use IIS currently and I was wondering if we can set up the advanced hosting through that instead of PM2

**Reply by charlesabarnes** [reply]
Pm2 is not required. Any runner for node.js apps work pm2 is the most popular.  

I am unsure why this is occuring. I will try to set this up on a windows box this weekend and see if I run into the same stuff. I will let you know.

**Reply by ekrocks** [reply]
Hello, were you able to get any of the similar issue on your windows box this weekend?

**Reply by charlesabarnes** [reply]
No I was not able to.  It was running as expected.  The piece I don't understand is the specific HTTP error you are receiving.  I am wondering if its a firewall or network issue of sorts.  I am still working to try to reproduce

**Reply by charlesabarnes** [reply]
Here's a writeup of the steps taken for this to get it working on my build machine
```
git clone https://github.com/bullhorn/career-portal
cd career-portal
nvm use 10.12.0 #node version I am running for career portal
npm install
npm run build
vim dist/browser/app.json  #add corp token and swimlane
pm2 start dist/server.js  #This step is what needs to be updated for the directory issue
```

**Reply by charlesabarnes** [reply]
These steps were done in git bash hence the forward slashes and the `nvm` command is a utility I downloaded to manage different node versions on windows

**Reply by ekrocks** [reply]
I will try reinstalling the necessary dependencies and try it again.

**Reply by ekrocks** [reply]
I wonder if it is a problem with the node version. Currently I have been using node version 14.17.6 which cannot use any node-sass <4.14 version. When I tried reinstalling the dependencies, I realized everything seemed to have installed correctly but node-sass. 
```> node-sass@4.10.0 install C:\Users\USERS\source\reposVSC\projectname\node_modules\@angular-devkit\build-angular\node_modules\node-sass
> node scripts/install.js

Downloading binary from https://github.com/sass/node-sass/releases/download/v4.10.0/win32-x64-83_binding.node
Cannot download "https://github.com/sass/node-sass/releases/download/v4.10.0/win32-x64-83_binding.node":
```
I seem to be getting this error even though I run the command ```npm install node-sass@4.14```. Do I need to edit package.json file or anything that will let me install the correct node-sass version? (Hoping that this might be the cause of the issue). 

Would it be better if I just tried installing the nvm and specify the node version like you have done above?

UPDATE: I have followed all your steps above (cloned a new project, installed & configured nvm to set node version = 10.12.0) but I am still getting the 
```
0|server  | { Error: connect ETIMEDOUT IPADDRESS:443
0|server  |     at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1113:14)
0|server  |   errno: 'ETIMEDOUT',
0|server  |   code: 'ETIMEDOUT',
0|server  |   syscall: 'connect',
0|server  |   address: 'ADDRESS',
0|server  |   port: 443 }
```
during ```npm install`` I noticed these warnings: 
```
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@2.3.2 (node_modules\watchpack\node_modules\fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.3.2: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.13 (node_modules\fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.13: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})
```
I searched up fsevents are they are responsible for directory updates? Would this cause anything?

**Reply by charlesabarnes** [reply]
Looks like it's trying to use the GitHub registry.  Can you switch to using the npm one for that?

**Reply by charlesabarnes** [reply]
Oh I see your edit

**Reply by ekrocks** [reply]
The more I search, the more arrows point to the proxy not being set up with pm2. Not sure if pm2 supports proxy but I will look more into that direction.

**Reply by ekrocks** [reply]
I may have found where the error is getting thrown. I have tried running the server through ```npm start ``` and typing ```localhost:4000/sitemap```, it actually outputs: 
```
C:\Users\USER\source\reposVSC\career-portal\dist\server.js:398
                        throw error;
                        ^
```

Which points to this line in ```server.js```
```
Zone.prototype.runTask = function (task, applyThis, applyArgs) {
            if (task.zone != this) {
                throw new Error('A task can only be run in the zone of creation! (Creation: ' +
                    (task.zone || NO_ZONE).name + '; Execution: ' + this.name + ')');
            }
            // https://github.com/angular/zone.js/issues/778, sometimes eventTask
            // will run in notScheduled(canceled) state, we should not try to
            // run such kind of task but just return
            if (task.state === notScheduled && (task.type === eventTask || task.type === macroTask)) {
                return;
            }
            var reEntryGuard = task.state != running;
            reEntryGuard && task._transitionTo(running, scheduled);
            task.runCount++;
            var previousTask = _currentTask;
            _currentTask = task;
            _currentZoneFrame = { parent: _currentZoneFrame, zone: this };
            try {
                if (task.type == macroTask && task.data && !task.data.isPeriodic) {
                    task.cancelFn = undefined;
                }
                try {
                    return this._zoneDelegate.invokeTask(this, task, applyThis, applyArgs);
                }
                catch (error) { <--- This is line 398
                    if (this._zoneDelegate.handleError(this, error)) {
                        throw error;
                    }
                }
            }
```

The ```catch(error)``` is getting caught and its stopping the process of creating the sitemap. Any ideas on this?

**Reply by charlesabarnes** [reply]
no idea right now.  Think you can add a console.log to the catch to hopefully shine some more light?

**Reply by ekrocks** [reply]
It was a problem with proxy settings, I was able to make it work when I made it go past the proxy. Thank you for all the help
